### PR TITLE
fix: popularity algorithm w/o stalling

### PR DESCRIPTION
**Pull request by blaggacao**
- this popularity algorithm can't end up stalling

### CODE REVIEW

**Comment by blaggacao** [comment]
Converting to draft since it looks like the topo-sort algorithm isn't stable. Solution: add a alphanumerical sort on packages with same popularity.

**Comment by blaggacao** [comment]
> Converting to draft since it looks like the topo-sort algorithm isn't stable. Solution: add a alphanumerical sort on packages with same popularity.

Implemented in https://github.com/nlewo/nix2container/pull/30/commits/a2977766cf63cc062ea93bbb510e444a7181862b

**Comment by blaggacao** [comment]
> Converting to draft since it looks like the topo-sort algorithm isn't stable. Solution: add a alphanumerical sort on packages with same popularity.

Implemented via https://github.com/nlewo/nix2container/pull/30/commits/faa611d31c5c5e902c58b2f8d9c5cdc4c489daff

File: closure/layered.go [hunk 1]
```diff
@@ -1,130 +1,52 @@
...
-func buildGraphNode(storepath Storepath, storepaths []Storepath) (node Node) {
-	node.Path = storepath.Path
-	node.Popularities = make(map[string]int)
-	for _, r := range storepath.References {
-		if r == node.Path {
-			continue
+// SortedPathsByPopularity sorts storepaths by path popularity. It uses the algorithm described in
+// https://grahamc.com/blog/nix-and-layered-docker-images
+func SortedPathsByPopularity(storepaths []Storepath) ([]string, error) {
+	g := simple.NewDirectedGraph()
```
Comment (nlewo) [inline closure/layered.go:12]: I would prefer to move this part in a dedicated function to indicate it builds the graph.

**Review by nlewo — COMMENTED** [review]
Yeah, that's pretty cool!

Could you add a test case for the infinite recursion your observed?

(i still need to review the `Score` function)

**Comment by nlewo** [comment]
`Score` function LGTM!

Thx for this reimplementation ;)

File: closure/layered.go (hunk 1 above)
Comment (blaggacao) [reply closure/layered.go:12]: doing/done

**Comment by blaggacao** [comment]
> Could you add a test case for the infinite recursion your observed?

It was a graph with 780 dependencies, hence I shied away from finding the root cause.

In principle, an iterator approach works because it locks on a cycle, instead of recursing

File: closure/layered.go (hunk 1 above)
Comment (blaggacao) [reply closure/layered.go:10]: https://github.com/nlewo/nix2container/pull/30/commits/3f335ade6a2ab1e0f8f4a8be5d0f3070a14833b5

**Comment by blaggacao** [comment]
Maybe it hasn't been an infinite recursion, but just an inefficient algorithm that would have stalled?

I ran the original build that run out of memory and cpu for a couple of seconds with a modified n2c with extra logs:
```
layers.json> INFO[0027] Path 2 /nix/store/1my9xr1s1nfjmqwyi46pzdrvny7hm66x-zlib-1.2.11-dev
layers.json> INFO[0027] Path 1 (counter 1724) /nix/store/1my9xr1s1nfjmqwyi46pzdrvny7hm66x-zlib-1.2.11-dev
layers.json> INFO[0027] Path 2 /nix/store/bqjh8dc6m5plswxrmmn9x4cpxrrw94lv-zlib-1.2.11
layers.json> INFO[0027] Path 1 (counter 6893) /nix/store/bqjh8dc6m5plswxrmmn9x4cpxrrw94lv-zlib-1.2.11
layers.json> INFO[0027] Path 2 /nix/store/4s21k8k7p1mfik0b33r2spq5hq7774k1-glibc-2.33-108
layers.json> INFO[0027] Path 1 (counter 72367) /nix/store/4s21k8k7p1mfik0b33r2spq5hq7774k1-glibc-2.33-108
layers.json> INFO[0027] Path 2 /nix/store/w2id1hwv4vv7hvp4slgsyrydrjbfqdxc-libidn2-2.3.2
layers.json> INFO[0027] Path 1 (counter 72367) /nix/store/w2id1hwv4vv7hvp4slgsyrydrjbfqdxc-libidn2-2.3.2
layers.json> INFO[0027] Path 2 /nix/store/8ckxc8biqqfdwyhr0w70jgrcb4h7a4y5-libunistring-0.9.10
layers.json> INFO[0027] Path 1 (counter 1724) /nix/store/2jm1a1sj36zq2lywqd3ybl11ypqi78b2-nodejs-14.19.0
layers.json> INFO[0027] Path 2 /nix/store/2aprvflf41z1dfma87mzd3md51xm2g71-openssl-1.1.1m
layers.json> INFO[0027] Path 1 (counter 6893) /nix/store/2aprvflf41z1dfma87mzd3md51xm2g71-openssl-1.1.1m
layers.json> INFO[0027] Path 2 /nix/store/4s21k8k7p1mfik0b33r2spq5hq7774k1-glibc-2.33-108
layers.json> INFO[0027] Path 1 (counter 72368) /nix/store/4s21k8k7p1mfik0b33r2spq5hq7774k1-glibc-2.33-108
layers.json> INFO[0027] Path 2 /nix/store/w2id1hwv4vv7hvp4slgsyrydrjbfqdxc-libidn2-2.3.2
layers.json> INFO[0027] Path 1 (counter 72368) /nix/store/w2id1hwv4vv7hvp4slgsyrydrjbfqdxc-libidn2-2.3.2
layers.json> INFO[0027] Path 2 /nix/store/8ckxc8biqqfdwyhr0w70jgrcb4h7a4y5-libunistring-0.9.10
layers.json> INFO[0027] Path 1 (counter 1724) /nix/store/2jm1a1sj36zq2lywqd3ybl11ypqi78b2-nodejs-14.19.0
layers.json> INFO[0027] Path 2 /nix/store/31iz8rdckish6i7hqdqw3870mb48pgjr-libuv-1.43.0
layers.json> INFO[0027] Path 1 (counter 1724) /nix/store/31iz8rdckish6i7hqdqw3870mb48pgjr-libuv-1.43.0
layers.json> INFO[0027] Path 2 /nix/store/4s21k8k7p1mfik0b33r2spq5hq7774k1-glibc-2.33-108
layers.json> INFO[0027] Path 1 (counter 72369) /nix/store/4s21k8k7p1mfik0b33r2spq5hq7774k1-glibc-2.33-108
layers.json> INFO[0027] Path 2 /nix/store/w2id1hwv4vv7hvp4slgsyrydrjbfqdxc-libidn2-2.3.2
layers.json> INFO[0027] Path 1 (counter 72369) /nix/store/w2id1hwv4vv7hvp4slgsyrydrjbfqdxc-libidn2-2.3.2
layers.json> INFO[0027] Path 2 /nix/store/8ckxc8biqqfdwyhr0w70jgrcb4h7a4y5-libunistring-0.9.10
layers.json> INFO[0027] Path 1 (counter 1724) /nix/store/2jm1a1sj36zq2lywqd3ybl11ypqi78b2-nodejs-14.19.0
layers.json> INFO[0027] Path 2 /nix/store/4nmqxajzaf60yjribkgvj5j54x9yvr1r-bash-5.1-p12
layers.json> INFO[0027] Path 1 (counter 8616) /nix/store/4nmqxajzaf60yjribkgvj5j54x9yvr1r-bash-5.1-p12
layers.json> INFO[0027] Path 2 /nix/store/4s21k8k7p1mfik0b33r2spq5hq7774k1-glibc-2.33-108
layers.json> INFO[0027] Path 1 (counter 72370) /nix/store/4s21k8k7p1mfik0b33r2spq5hq7774k1-glibc-2.33-108
layers.json> INFO[0027] Path 2 /nix/store/w2id1hwv4vv7hvp4slgsyrydrjbfqdxc-libidn2-2.3.2
```

After only a couple of seconds we already can see `(counter 72370)` recursions (sic!) passing by `glibc`.

**Review by nlewo — APPROVED** [review]
Cool, thanks you!

**Pull request merged**